spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java

```diff
@@ -0,0 +1,377 @@
+
+    // Check that different widths can be used
+    Assert.assertEquals((byte) -2, scalarSql("SELECT system.truncate(2, -1Y)"));
+
+    // Check that tinyint types are allowed for the width
+    Assert.assertEquals((byte) 0, scalarSql("SELECT system.truncate(5Y, 1Y)"));
+
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS tinyint))"));
+  }
+
+  @Test
+  public void testTruncateSmallInt() {
+    Assert.assertEquals((short) 0, scalarSql("SELECT system.truncate(10, 0S)"));
+    Assert.assertEquals((short) 0, scalarSql("SELECT system.truncate(10, 1S)"));
+    Assert.assertEquals((short) 0, scalarSql("SELECT system.truncate(10, 5S)"));
+    Assert.assertEquals((short) 0, scalarSql("SELECT system.truncate(10, 9S)"));
+    Assert.assertEquals((short) 10, scalarSql("SELECT system.truncate(10, 10S)"));
+    Assert.assertEquals((short) 10, scalarSql("SELECT system.truncate(10, 11S)"));
+    Assert.assertEquals((short) -10, scalarSql("SELECT system.truncate(10, -1S)"));
+    Assert.assertEquals((short) -10, scalarSql("SELECT system.truncate(10, -5S)"));
+    Assert.assertEquals((short) -10, scalarSql("SELECT system.truncate(10, -10S)"));
+    Assert.assertEquals((short) -20, scalarSql("SELECT system.truncate(10, -11S)"));
+
+    // Check that different widths can be used
+    Assert.assertEquals((short) -2, scalarSql("SELECT system.truncate(2, -1S)"));
+
+    // Check that short types are allowed for the width
+    Assert.assertEquals((short) 0, scalarSql("SELECT system.truncate(5S, 1S)"));
+
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS smallint))"));
+  }
+
+  @Test
+  public void testTruncateInt() {
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 0));
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 1));
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 5));
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(10, %d)", 9));
+    Assert.assertEquals(10, scalarSql("SELECT system.truncate(10, %d)", 10));
+    Assert.assertEquals(10, scalarSql("SELECT system.truncate(10, %d)", 11));
+    Assert.assertEquals(-10, scalarSql("SELECT system.truncate(10, %d)", -1));
+    Assert.assertEquals(-10, scalarSql("SELECT system.truncate(10, %d)", -5));
+    Assert.assertEquals(-10, scalarSql("SELECT system.truncate(10, %d)", -10));
+    Assert.assertEquals(-20, scalarSql("SELECT system.truncate(10, %d)", -11));
+
+    // Check that different widths can be used
+    Assert.assertEquals(-2, scalarSql("SELECT system.truncate(2, %d)", -1));
+    Assert.assertEquals(0, scalarSql("SELECT system.truncate(300, %d)", 1));
+
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS int))"));
```

Comment:
Does this call `produceResult` or `invoke`?

Reply:
This calls `invoke`. For `invoke`, if any primitive input is `null`, `null` is returned automatically when using the magic functions -- for non-primitives like UTF8String, we still need to check.

Here's the query plan to show it:
```
    == Parsed Logical Plan ==
    'Project [unresolvedalias('system.truncate(2, cast(null as bigint)), None)]
    +- OneRowRelation
    == Analyzed Logical Plan ==
    staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(NULL as bigint), IntegerType, LongType, false, true, true): bigint
    Project [staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(null as bigint), IntegerType, LongType, false, true, true) AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(NULL as bigint), IntegerType, LongType, false, true, true)#55L]
    +- OneRowRelation
    == Optimized Logical Plan ==
    Project [null AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(NULL as bigint), IntegerType, LongType, false, true, true)#55L]
    +- OneRowRelation
    == Physical Plan ==
    *(1) Project [null AS staticinvoke(class org.apache.iceberg.spark.functions.TruncateFunction$TruncateBigInt, LongType, invoke, 2, cast(NULL as bigint), IntegerType, LongType, false, true, true)#55L]
    +- *(1) Scan OneRowRelation[]
```